### PR TITLE
Ported STL10 dataset's tests to new test framework

### PR DESCRIPTION
**Pull request by prabhat00155**
Discussion: https://github.com/pytorch/vision/issues/3531

### CODE REVIEW

File: test/test_datasets.py [hunk 1]
```diff
@@ -141,76 +145,68 @@ def test_places365_repr_smoke(self):
...
-        for split in ('train', 'train+unlabeled', 'unlabeled', 'test'):
-            with self.mocked_dataset(split=split) as (dataset, data):
-                num_images = sum([data["num_images_in_split"][part] for part in split.split("+")])
-                self.generic_classification_dataset_test(dataset, num_images=num_images)
-
-    def test_folds(self):
-        for fold in range(10):
-            with self.mocked_dataset(split="train", folds=fold) as (dataset, data):
-                num_images = data["num_images_in_folds"][fold]
-                self.assertEqual(len(dataset), num_images)
```
Comment (pmeier) [inline test/test_datasets.py:173]: This test is missing for now. It should test that the number of images in a fold is correctly picked up.

File: test/test_datasets.py [hunk 2]
```diff
@@ -141,76 +145,68 @@ def test_places365_repr_smoke(self):
...
-
-    def test_invalid_folds1(self):
-        with self.assertRaises(ValueError):
-            with self.mocked_dataset(folds=10):
-                pass
-
-    def test_invalid_folds2(self):
-        with self.assertRaises(ValueError):
-            with self.mocked_dataset(folds="0"):
-                pass
```
Comment (pmeier) [inline test/test_datasets.py:183]: These tests are missing for now. They test if an error is raised in case a "wrong" `fold` is passed.

File: test/test_datasets.py [hunk 3]
```diff
@@ -141,76 +145,68 @@ def test_places365_repr_smoke(self):
...
+            with open(os.path.join(root, "fold_indices.txt"), "w") as fh:
+                for fold in range(num_folds):
+                    line = " ".join([str(idx) for idx in range(offset, offset + fold + 1)])
+                    fh.write(f"{line}\n")
+                    offset += fold + 1
 
-    def test_unlabeled(self):
-        with self.mocked_dataset(split="unlabeled") as (dataset, _):
-            labels = [dataset[idx][1] for idx in range(len(dataset))]
-            self.assertTrue(all([label == -1 for label in labels]))
```
Comment (pmeier) [inline test/test_datasets.py:204]: This test is missing for now. It should test that the labels in the `"unlabeled"` split always are `-1`.

File: test/test_datasets.py [hunk 4]
```diff
@@ -141,76 +145,68 @@ def test_places365_repr_smoke(self):
...
-    def test_invalid_folds2(self):
-        with self.assertRaises(ValueError):
-            with self.mocked_dataset(folds="0"):
-                pass
-
-    def test_transforms(self):
-        expected_image = "image"
-        expected_target = "target"
+    def inject_fake_data(self, tmpdir, config):
+        def make_binary_file(num_elements, root, name):
```
Comment (pmeier) [inline test/test_datasets.py:154]: Is there a reason to define these functions locally instead as instance methods? If not, I would prefer the latter for readability.

File: test/test_datasets.py [hunk 5]
```diff
@@ -141,76 +145,68 @@ def test_places365_repr_smoke(self):
...
+            make_tar(root, name, name, compression="gz")
+
+        archive_name = "stl10_binary"
+        archive_folder = os.path.join(tmpdir, archive_name)
+        os.mkdir(archive_folder)
+
+        num_images_in_split = make_train_files(archive_folder)
+        num_images_in_split.update(make_test_files(archive_folder))
+        make_class_names_file(archive_folder)
+        make_archive(tmpdir, archive_name)
```
Comment (pmeier) [inline test/test_datasets.py:207]: Following the docs for `inject_fake_data`

https://github.com/pytorch/vision/blob/eee424922e998dddcaabe927e823ec515aa2927b/test/datasets_utils.py#L308-L310

we should not create an archive here, but rather provide the data in the format after the extraction.

**Review by pmeier — CHANGES_REQUESTED** [review]
Hey @prabhat00155 thanks for the PR. I have a few comments below. In general, we should make sure that we don't reduce coverage while porting dataset tests.

File: test/test_datasets.py (hunk 4 above)
Comment (NicolasHug) [reply test/test_datasets.py:154]: I think the main reason is that these functions are only ever used in `inject_fake_data`. Making them a method (even private) would hint that these may be used in more than one place. Also, they don't use or mutate `self` at all.

I tend to favor nested functions for this as well, although I noticed you had a different opinion already :p

These are just tests so no strong opinion from me anyway

Comment (pmeier) [reply test/test_datasets.py:154]: I wouldn't be against it if it is only a small function. But in this case these functions make up most of the body of `inject_fake_data`. I actually had to scroll down from the definition to reach the code that is actually executed. 

> Making them a method (even private) would hint that these may be used in more than one place.

Not necessarily. You can simply factor out code to enhance the readability. For example look at the `download` method. We only call this once during instantiation, but it is super easy to read

https://github.com/pytorch/vision/blob/eee424922e998dddcaabe927e823ec515aa2927b/torchvision/datasets/stl10.py#L62-L63

without having a massive constructor.

> Also, they don't use or mutate self at all.

I don't mind if they are static.

File: test/test_datasets.py [hunk 6]
```diff
@@ -141,76 +141,86 @@ def test_places365_repr_smoke(self):
             self.assertIsInstance(repr(dataset), str)
 
 
-class STL10Tester(DatasetTestcase):
-    @contextlib.contextmanager
-    def mocked_root(self):
-        with stl10_root() as (root, data):
-            yield root, data
+class STL10Tester(datasets_utils.ImageDatasetTestCase):
```
Comment (pmeier) [inline test/test_datasets.py:144]: To keep it in sync with the other test cases:

```suggestion
class STL10TestCase(datasets_utils.ImageDatasetTestCase):
```

File: test/test_datasets.py [hunk 7]
```diff
@@ -141,76 +141,86 @@ def test_places365_repr_smoke(self):
...
-            if pre_extract:
-                utils.extract_archive(os.path.join(root, data["archive"]))
-            dataset = torchvision.datasets.STL10(root, download=download, **kwargs)
-            yield dataset, data
-
-    def test_not_found(self):
-        with self.assertRaises(RuntimeError):
-            with self.mocked_dataset(download=False):
-                pass
+    @staticmethod
```
Comment (pmeier) [inline test/test_datasets.py:149]: Why are some methods static, but others are not? IMO we should do either one or the other.

File: test/test_datasets.py [hunk 8]
```diff
@@ -141,76 +141,86 @@ def test_places365_repr_smoke(self):
...
 
-    def test_invalid_folds1(self):
-        with self.assertRaises(ValueError):
-            with self.mocked_dataset(folds=10):
-                pass
+    @staticmethod
+    def _make_class_names_file(root, name="class_names.txt"):
+        class_names = ("airplane", "bird")
+        with open(os.path.join(root, name), "w") as fh:
+            for cname in class_names:
```
Comment (pmeier) [inline test/test_datasets.py:164]: Nit:

```suggestion
        with open(os.path.join(root, name), "w") as fh:
            for cname in ("airplane", "bird"):
```

**Review by pmeier — APPROVED** [review]
Few nitpicks left, but otherwise LGTM! Thanks @prabhat00155.
